spinserver.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.37.1.
.TH SPINSERVER "1" "April 2012" "spinserver " "User Commands"
.SH NAME
spinserver \- server for the SPIN Framework
.SH SYNOPSIS
spinserver [options] <scene-to-load.xml>
.SH DESCRIPTION
=====================================================
SpatOSC (version 0.3.1)
[Translators]:: 0 translators:
[Scene]:: connectFilter = .*
[Scene]:: 0 listeners:
[Scene]:: 0 sources:
[Scene]:: 0 connections:
.SH OPTIONS
.TP
\fB\-\-disable\-auto\-cleanup\fR
Disables the auto\-cleanup of user nodes if they stop sending
ping messages.
.TP
\fB\-\-disable\-discovery\fR
Disable multicast discovery services.
.TP
\fB\-\-recv\-tcp\-msg\fR <port>
Set the port where we listen for subscription requests from
clients. Clients may also send scene events to this port is
they desire reliability. (Default: 54322)
.TP
\fB\-\-recv\-udp\-msg\fR <host> <port>
Set the address/port for listening to UDP messages from
clients. This argument may be repeated for multiple
multicast groups and/or ports (Default: 239.0.0.1 54323)
.TP
\fB\-\-scene\-id\fR <id>
Specify the id of the SPIN scene (Default: default)
.TP
\fB\-\-send\-udp\-msg\fR <host> <port>
Set the address/port for UDP multicast of scene events, or
this argument may be repeated for several unicast
connections (Default: 239.0.0.1 54324)
.TP
\fB\-\-send\-udp\-sync\fR <host> <port>
Set the address/port for timecode (sync) messages (Default:
239.0.0.1 54321)
.TP
\fB\-\-spatosc\fR <translator> <URL>
Enables SPIN's internal SpatOSC scene. Example: \fB\-\-spatosc\fR
BasicTranslator osc.tcp://localhost:18033
.TP
\fB\-\-ttl\fR <number>
Set the TTL (time to live) for multicast packets in order to
hop across routers (Default: 1)
.TP
\fB\-\-version\fR
Display the version number and exit.
.TP
\fB\-h\fR or \fB\-\-help\fR
Display this information
.PP
=====================================================
SpatOSC (version 0.3.1)
[Translators]:: 0 translators:
[Scene]:: connectFilter = .*
[Scene]:: 0 listeners:
[Scene]:: 0 sources:
[Scene]:: 0 connections:
0.3.13
.SH "REPORTING BUGS"
https://sourceforge.net/apps/trac/spinframework/
.SH "SEE ALSO"
spinviewer (1)


NEWS
SPIN Framework release notes
============================

2012-0?-?? : spinframework 0.3.14
---------------------------------
This release is in the development series 0.3.x that has an unstable API. 

New features:

* (none)

Bugs fixed:

* (none)


2012-03-07 : spinframework 0.3.12
---------------------------------
This release is in the development series 0.3.x that has an unstable API. 

Bugs fixed:

* Get rid of OSX-specific hack


2012-03-05 : spinframework 0.3.10
---------------------------------
This release is in the development series 0.3.x that has an unstable API. 

New features:

* added option of reading command line args from file: ~/.spinFramework/args
* added initial PointCloud node
* added GeometryNode
* xcode project
* SharedVideoTexture: allow set/get renderbin

Bugs fixed:

* fixed boost linking for spinserver and spinviewer app
* added missing boost libraries to Makefile.am
* fixed texture loading for Ubuntu using OSG 3.0.1 from package (by specifying the correct plugin path)
* added /usr/local/lib and /usr/local/lib64 to LDFLAGS to test on Ubuntu Precise
* switched to new boost.m4 script
* changed ifdefs in sharedvideonode to prevent errors if wrappers are generated in linux
* set video textures to play by default
* set OSG-specific library path in one_step_build.sh
* delete SharedVideoNode
* add Makefile.am for completion dir
* fixed orbit example's Makefile
* added boost-program-options library, since the newest python 2.6 on Lucid uses an updated libutil, which requires boost-program-options


2012-01-13 : spinframework 0.3.4
--------------------------------
This release is in the development series 0.3.x that has an unstable API. 

New features:

* Enclose everything in the spin namespace
* four-screen cube
* dome distorsion to camera config files
* Library renamed to libspin-0.3
* add pkg-config file
* update spatosc arguments, since spatosc takes OSC urls (ie, osc.udp://host:port) instead of host and port arguments separately
* changed spinserver's txAddr to a list of addresses, so we can use multiple UDP unicast connections instead of multicast (important on wacky networks that don't permit multicast)
* big restructuring of command-line arguments, allowing for manual override of everything
* added command-line options for specifying the TCP port for both the client and server contexts
* added camconfig .xml files
* added TTL for multicast packets, and updated some wrappers
* use cppintrospection 3.0
* document ShapeNode, TextureStateSetFinder, spinApp, PointerNodeActionAdapter, NodeSearcher, Fog
* removed ^M characters
* list all spin nodes at startup
* add Menu3D and AttractorNode
* add some features to TextNode
* added ReporterNode
* added bounce and collide events for ConstraintsNode

Bugs fixed:

* TIMECODE CHANGE: now using double precision milliseconds
* fixed bug when compiling spin without spatosc
* fixed autoconf macro for python so that it detects python 2.7 properly (Natty)
* disallowed spinserver to start with automatic port if user manually specifies a port that is already taken. Instead, we quit with an error message
* fixed nodelist sending to multiple clients (but only for UDP. Still need to fix TCPvi src/spin/spinApp.cpp )
* fixed buffer accumulation bug with SharedVideoTexture (un-assigned textures were not updating, and thus, not eating frames)
* using glibtool instead of libtool on OSX
* fixed a few old osgIntrospection keyworkds
* needed to set a function as an extern "C"
* fixed memleak spinBaseContext
* ignoring lighting for ShapeNodes with statesets
* fixed test - not launching a listener
* ensuring that QuatToEuler reports angles in range of -PI to PI
* progress with signal handling for editor, and ported the TTL stuff from svn version of spin
* more forward declaration
* removed check for fmod
* ViewerManipulator: fixed endless loop
* fixed saving of statesets in xml, and moved towards proper .app bundles for OSX
* Changed spinviewer to use pollUpdates() method, sharing a single thread with OSG
* spinviewer::run should return 0
* Use ports and addresses in spinDefaults. Update multicast group to 239.0.0.1


2010-07-22 : spinframework 0.3.2
--------------------------------
This release is in the development series 0.3.x that has an unstable API. It introduces
many bug fixes and new features. We changed the UDP multicast group to 226.0.0.1 for the 
info and scene message channels. We fixed a lot of bugs regarding the shared video texture
using the Scenic milhouse video streaming utility. 

New features:

* Added a TCP channel for important messages to the spinserver
* Renamed spinserver and spinviewer to lowercase
* Implemented the --version option in spinserver and spinviewer
* Created man pages for spinserver and spinviewer
* Introducing the server side Python scripting
* big refactoring of spinContext
* updated multicast group to 226.0.0.1
* created RELEASE, AUTHORS and NEWS files
* using POSIX/GNU-style options for spinserver and spinviewer
* Using /SPIN/__user__ for user subscription
* Using left-hand coordinates for both rotations and translations
* don't need to configure with --enable-scenic, if it's detected it will be used 

Bugs fixed:

* fixed sync signal to be double float in ms
* updates to autoconf for Ubuntu Lucid
* fixed `make dist`
* updated glut m4
* made spin context functions non global
* got rid of -Wstrict-prototypes in python cflags
* comment out fmod
* added -Wfatal-errors to flags
* don't use using namespace std
* moved python scripts into Resources directory
* fixed rm rule in osgwrappers
* fixed spinviewer cleanup problem by using a joinable thread rather than a detached thread
* fixed the way that statesets are cleared (clearStates) message 
* added more aggressive interrupt handler 
* #12 Removed the src/deprecated directory from the trunk
* added non-blocking keyboard input to example so that ctrl-c works 
* fixed memory leaks in shared video texture 

src/spin/spinServer.cpp
// -----------------------------------------------------------------------------
// |    ___  ___  _  _ _     ___                                        _      |
// |   / __>| . \| || \ |   | __>_ _  ___ ._ _ _  ___  _ _ _  ___  _ _ | |__   |
// |   \__ \|  _/| ||   |   | _>| '_><_> || ' ' |/ ._>| | | |/ . \| '_>| / /   |
// |   <___/|_|  |_||_\_|   |_| |_|  <___||_|_|_|\___.|__/_/ \___/|_|  |_\_\   |
// |                                                                           |
// |---------------------------------------------------------------------------|
//
// http://spinframework.sourceforge.net
// Copyright (C) 2009 Mike Wozniewski, Zack Settel
//
// Developed/Maintained by:
//    Mike Wozniewski (http://www.mikewoz.com)
//    Zack Settel (http://www.sheefa.net/zack)
//
// Principle Partners:
//    Shared Reality Lab, McGill University (http://www.cim.mcgill.ca/sre)
//    La Societe des Arts Technologiques (http://www.sat.qc.ca)
//
// Funding by:
//    NSERC/Canada Council for the Arts - New Media Initiative
//    Heritage Canada
//    Ministere du Developpement economique, de l'Innovation et de l'Exportation
//
// -----------------------------------------------------------------------------
//  This file is part of the SPIN Framework.
//
//  SPIN Framework is free software: you can redistribute it and/or modify
//  it under the terms of the GNU Lesser General Public License as published by
//  the Free Software Foundation, either version 3 of the License, or
//  (at your option) any later version.
//
//  SPIN Framework is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU Lesser General Public License for more details.
//
//  You should have received a copy of the GNU Lesser General Public License
//  along with SPIN Framework. If not, see <http://www.gnu.org/licenses/>.
// -----------------------------------------------------------------------------

#include <string>
#include <iostream>
#include <boost/filesystem.hpp>

#include <osgViewer/CompositeViewer>
#include <osg/ArgumentParser>
#include <osgDB/ReadFile>
#include <osg/Timer>


#include "SceneManager.h"
#include "spinApp.h"
#include "spinServerContext.h"
#include "spinLog.h"
#include "config.h"

#ifdef HAVE_POCO
#include "Poco/Net/HTTPServer.h"
#include "Poco/Net/HTTPRequestHandler.h"
#include "Poco/Net/HTTPRequestHandlerFactory.h"
#include "Poco/Net/HTTPServerParams.h"
#include "Poco/Net/HTTPServerRequest.h"
#include "Poco/Net/HTTPServerResponse.h"
#include "Poco/Net/HTTPServerParams.h"
#include "Poco/Net/HTMLForm.h"
#include "Poco/Net/PartHandler.h"
#include "Poco/Net/MessageHeader.h"
#include "Poco/Net/ServerSocket.h"
#include "Poco/CountingStream.h"
#include "Poco/NullStream.h"
#include "Poco/StreamCopier.h"
#include "Poco/Exception.h"
#include "Poco/Util/ServerApplication.h"
#include "Poco/Util/Option.h"
#include "Poco/Util/OptionSet.h"
#include "Poco/Util/HelpFormatter.h"
#endif

// *****************************************************************************
// *****************************************************************************
// *****************************************************************************
int main(int argc, char **argv)
{
	using namespace spin;
	spinServerContext server;
    
	// *************************************************************************
    // If no command line arguments were passed, check if there is an args file
    // at ~/.spinFramework/args and override argc and argv with those:
    std::vector<char*> newArgs = getUserArgs();
    if ((argc==1) && (newArgs.size() > 1))
    {
        newArgs.insert(newArgs.begin(), argv[0]);
        argc = (int)newArgs.size()-1;
        argv = &newArgs[0];
    }
    
	// *************************************************************************
	// PARSE ARGUMENTS::

	osg::ArgumentParser arguments(&argc,argv);
	
    // set up the usage document, which a user can acess with -h or --help
	arguments.getApplicationUsage()->setDescription(arguments.getApplicationName()+" is a server for the SPIN Framework.");
	arguments.getApplicationUsage()->setCommandLineUsage(arguments.getApplicationName()+" [options] <scene-to-load.xml>");
    
    // add generic spin arguments (for address/port setup, scene-id, etc) 
    server.addCommandLineOptions(&arguments);

    // arguments specific to spinserver:
    arguments.getApplicationUsage()->addCommandLineOption("--disable-discovery", "Disable multicast discovery services.");

	// PARSE ARGS:

    // parse generic args:
    if (!server.parseCommandLineOptions(&arguments))
        return 0;

	// any option left unread are converted into errors to write out later
	arguments.reportRemainingOptionsAsUnrecognized();

	
	// *************************************************************************
	// remaining arguments are assumed to be scene elements to be loaded (.xml)
	
	std::vector<std::string>::iterator arg;
	for (int i=1; i<argc; i++) //arg=arguments.begin(); arg!=arguments.end(); ++arg)
	{
		if (arguments[i][0]!='-')
		{
			// not an option so assume string is a filename
			spinApp::Instance().sceneManager->loadXML(arguments[i]);
		} else i++;
	}
	 	
	// any option left unread are converted into errors
	//arguments.reportRemainingOptionsAsUnrecognized();
	
	if (arguments.errors())
	{
		arguments.writeErrorMessages(std::cout);
		return 1;
	}

	// *************************************************************************
	// start spin:
	server.start();


	// *************************************************************************
    // launch webserver:
    
#ifdef HAVE_POCO
    //unsigned short port = 9980;
			
    // set-up a server socket
	//Poco::Net::ServerSocket svs(port);
#endif    

	// *************************************************************************
	// send a userRefresh message:
    SCENE_MSG("s", "userRefresh");
	
	// *************************************************************************
	// loop:
    std::cout << "\nspinserver is READY" << std::endl;
    try {	
        while (server.isRunning())
        {
            sleep(1);
            // loop until a quit message is received (TODO)
        }
    }
    catch (const std::exception &e)
    {
        std::cerr << "Got exception " << e.what() << std::endl;
        return 1;
    }

    usleep(100);
    std::cout << "spinserver exited normally." << std::endl;

    return 0;
}
